### PR TITLE
feat: ignore sudo error on ACK

### PR DESCRIPTION
**Pull request by albertandrejev**
In this PR fix for issue #2 from the audit report will be added.

In case of Sudo handler error state of the contract will be reverted and ACK as well as acknowledgement identifiers will be stored in the store. New module to get list of failed acks ids with contracts addresses will be added. Module name will be `contractmanager`

1. [Docs](https://github.com/neutron-org/neutron-docs/pull/30)
2. [Tests](https://github.com/neutron-org/neutron-integration-tests/pull/33)
3. [Relayer fix](https://github.com/neutron-org/neutron-query-relayer/pull/56)

More info: https://p2pvalidator.atlassian.net/browse/NTRN-253

### CODE REVIEW

File: x/contractmanager/client/cli/query.go
Comment (foxpy) [inline x/contractmanager/client/cli/query.go]: I would have deleted these commented imports

File: x/contractmanager/client/cli/query_failure.go
Comment (foxpy) [inline x/contractmanager/client/cli/query_failure.go:16]: Why do you define from 0 to 1 arguments? I don't see any argument handling code further.

File: x/contractmanager/client/cli/query_failure_test.go
Comment (foxpy) [inline x/contractmanager/client/cli/query_failure_test.go]: Seems like you could just delete it now.

**Review by foxpy — CHANGES_REQUESTED** [review]

**Review by foxpy — COMMENTED** [review]
Tested: this PR does NOT break integration tests.

File: x/contractmanager/client/cli/query_failure.go
Comment (albertandrejev) [reply x/contractmanager/client/cli/query_failure.go:16]: Forgot to remove it. Thanks for pointing

**Review by foxpy — APPROVED** [review]
utACK

**Comment by albertandrejev** [comment]
> utACK

?

**Comment by albertandrejev** [comment]
oops

File: proto/contractmanager/genesis.proto
Comment (sotnikov-s) [inline proto/contractmanager/genesis.proto]: re `ackID`, `ackType`, and `failureList` below: it's recommended to write field names in snake case:
https://developers.google.com/protocol-buffers/docs/style#message_and_field_names

File: proto/contractmanager/query.proto
Comment (sotnikov-s) [inline proto/contractmanager/query.proto]: I believe this query also requires pagination for a single address can have a great number of failures as well. also, since it returns a slice of failures, the name better be GetFailure**s**... or maybe GetFailuresByAddress?

File: proto/contractmanager/genesis.proto
Comment (sotnikov-s) [inline proto/contractmanager/genesis.proto:11]: please describe the `Failure` message itself

File: testutil/contractmanager/keeper/contractmanager.go
Comment (sotnikov-s) [inline testutil/contractmanager/keeper/contractmanager.go]: ContractmanagerKeeper -> Contract**M**anagerKeeper

File: testutil/contractmanager/keeper/contractmanager.go
Comment (sotnikov-s) [inline testutil/contractmanager/keeper/contractmanager.go]: ContractmanagerParams -> Contract**M**anagerParams

File: testutil/contractmanager/network/network.go
Comment (sotnikov-s) [inline testutil/contractmanager/network/network.go:30]: `configs ...network.Config` — is it a kind of general cosmos pattern? kinda strange taking into account the way we use this New and the returned value. I ask because I can see that we use the same network constructor in ICQ and ICTx modules whereas there are more clean (imo) ways to define optional parameters (e.g. `With...` methods, or several constructors, or even a `nil` value suppport)

this comment is just out of curiosity, an informal one

File: x/contractmanager/client/cli/query.go
Comment (zavgorodnii) [inline x/contractmanager/client/cli/query.go:27]: please remove all scaffolding mentions!

File: x/contractmanager/keeper/failure.go
Comment (zavgorodnii) [inline x/contractmanager/keeper/failure.go:37]: are we sure we want to be so silent about this?

File: x/contractmanager/client/cli/query.go
Comment (albertandrejev) [reply x/contractmanager/client/cli/query.go:27]: It is useful to have them because `ignite` is using it to alter files (to add new messages, queries etc).

File: x/contractmanager/client/cli/query_failure.go
Comment (sotnikov-s) [inline x/contractmanager/client/cli/query_failure.go]: shows a failures -> shows failures

File: x/contractmanager/keeper/failure.go
Comment (albertandrejev) [reply x/contractmanager/keeper/failure.go:37]: Yes, because our first ID is 0. We will return 0 in case if there is no any elements with specific contract address in the store.

File: x/transfer/keeper/keeper.go
Comment (zavgorodnii) [inline x/transfer/keeper/keeper.go]: private after public?

**Review by sotnikov-s — CHANGES_REQUESTED** [review]

File: x/contractmanager/keeper/sudo.go
Comment (sotnikov-s) [inline x/contractmanager/keeper/sudo.go:4]: the comment needs to be either removed or updated: it still says about the `sudo` package

File: x/contractmanager/types/keys.go
Comment (sotnikov-s) [inline x/contractmanager/types/keys.go]: maybe `"mem" + ModuleName`?

**Comment by oopcode** [comment]
[still looking at the recovery code]

don't we need the recovery logic also for ibc transfers?

File: x/interchainqueries/types/expected_keepers.go
Comment (sotnikov-s) [inline x/interchainqueries/types/expected_keepers.go:23]: I like this design of contract manager more than in the nearby PR. an idea: what do you think about naming the `contractmanager` module as simple `sudo`?

File: x/interchaintxs/keeper/keeper.go
Comment (sotnikov-s) [inline x/interchaintxs/keeper/keeper.go]: contractmanagerKeeper -> contract**M**anagerKeeper

File: x/interchainqueries/types/expected_keepers.go
Comment (albertandrejev) [reply x/interchainqueries/types/expected_keepers.go:23]: Nearby PR is not actual anymore. 
Not sure that I understand your idea about renaming, because current module provides not only sudo functionality.

File: x/interchaintxs/keeper/ibc_handlers.go
Comment (zavgorodnii) [inline x/interchaintxs/keeper/ibc_handlers.go]: i think that this uint overflow should be handled more explicitly, e.g.

```
var newLimit uint64
if GasReserve > gasLeft {
    newLimit = 0
} else {
    newLimit = gasLeft - GasReserve
}
```

File: x/interchainqueries/types/expected_keepers.go
Comment (sotnikov-s) [reply x/interchainqueries/types/expected_keepers.go:23]: the idea is simple: since the main idea of the ContractManager module is to store contracts sudo-related info, execute sudo calls to contracts and provide queries for getting this data, it's imo meaningful to rename it to Sudo: SudoKeeper, available via cli as `neutrond query sudo ...` etc. Maybe it even makes sense to split this module to two: `ContractManager` module with `HasContractInfo` and `AddContractFailure` methods, and a `Sudo` one with all the `Sudo...` methods. Out of this task scope. Wyt?

File: testutil/contractmanager/network/network.go
Comment (albertandrejev) [reply testutil/contractmanager/network/network.go:30]: This code is autogenerated by ignite so I guess it was created in this way to be more generic.

File: x/interchainqueries/types/expected_keepers.go
Comment (albertandrejev) [reply x/interchainqueries/types/expected_keepers.go:23]: Good idea, but I think that Sudo is too generic name and may be confusing. Also if we will split module into 2 different modules we will:
1. Move back to the previous state of the sudo helpers
2. We will have one module with ability to write to state and to query information from state, and another module without any interfaces to process messages or to query state. I mean this module will not have any ability to directly interact with cli on tx. Not sure that in this case we need to have another module for sudo, because previously we had this solution.

And current approach may help us to cover other contract related interactions in future.

File: x/contractmanager/client/cli/query_failure.go
Comment (sotnikov-s) [inline x/contractmanager/client/cli/query_failure.go:28]: btw is there a use of command execution with an empty address, i.e. of getting the whole failures happened in the chain? maybe the simple by-address-one is enough?

**Review by sotnikov-s — COMMENTED** [review]
LGTM, but I'll wait for integration tests before giving an approval

File: x/contractmanager/client/cli/query_failure.go
Comment (albertandrejev) [reply x/contractmanager/client/cli/query_failure.go:28]: At least you can. Do not see any issue with it.

File: proto/contractmanager/query.proto
Comment (swelf19) [inline proto/contractmanager/query.proto]: i do not see any difference between `AllFailures` and `Failures`. And code is the same for both methods. Method named `AllFailures` should at least omit the filtering with address params

File: x/interchaintxs/keeper/ibc_handlers.go
Comment (swelf19) [inline x/interchaintxs/keeper/ibc_handlers.go:17]: lines 16-57 the same in the transfer wrapper and interchain txs, is it possible to reuse the common code?

Comment (albertandrejev) [reply x/interchaintxs/keeper/ibc_handlers.go:17]: This is the problem, I want to have modules to be independent as possible. In order to reuse common code we need to introduce new entity, that will be used between different modules, but I do not want to break this modularity.

File: proto/contractmanager/query.proto
Comment (albertandrejev) [reply proto/contractmanager/query.proto]: the difference in the `address` parameter. But we need to have two grpc paths, one with address parameter another without. I am not sure that if I remove it (I mean second `rpc` endpoint) it will work properly with and without `address` parameter.

File: x/interchaintxs/keeper/ibc_handlers.go
Comment (zavgorodnii) [reply x/interchaintxs/keeper/ibc_handlers.go:17]: yup, we discussed it with Albert; let's leave it as is

File: proto/contractmanager/query.proto
Comment (swelf19) [reply proto/contractmanager/query.proto]: I got the idea, i mean there is no difference which method to call via (g)rpc, both takes the same arguments(paging, address), both executes the same code. What do you think to swap names `func (k Keeper) Failures` and `func (k Keeper) AllFailures`, its strange, when the more generic method `Failures` calls the more specific `AllFailures` and passes the filter arguments.

I've removed `AllFailures` from proto file, regenerated the code, and `Failures` works properly with empty `address`, but trailing slash is required.

Comment (albertandrejev) [reply proto/contractmanager/query.proto]: Tested. If I will remove one of the endpoints then other will stop working. It means if I will remove `AllFailures` rpc, then it will be longer not possible to get list of all failures. Same with other, in this case it will not be possible to get `Failures` for specific address.

Comment (albertandrejev) [reply proto/contractmanager/query.proto]: I tried it without slash :)

**Review by zavgorodnii — APPROVED** [review]

**Review by sotnikov-s — APPROVED** [review]

**Review by pr0n00gler — APPROVED** [review]

**Pull request merged**
